go/cmd/dolt/commands/sql_statement_scanner.go

```diff
@@ -54,89 +40,263 @@ const (
+	numConsecutiveDelimiterMatches int  // the consecutive number of characters that have been matched to the delimiter
+	statementStartLine             int
+}
+
+func (s *streamScanner) Scan() bool {
+	// truncate last query
+	s.truncate()
+	s.resetState()
+
+	if s.i >= s.fill {
+		// initialize buffer
+		if err := s.read(); err != nil {
+			s.err = err
+			return false
+		}
+	}
+
+	if s.isEOF || s.i == s.fill {
+		// no token
+		return false
+	}
+
+	// discard leading whitespace
+	for ; unicode.IsSpace(rune(s.buf[s.i])); s.i++ {
+		if s.buf[s.i] == '\n' {
+			s.lineNum++
+		}
+		if s.i >= s.fill {
+			if err := s.read(); err != nil {
+				s.err = err
+				return false
+			}
+		}
+	}
+	s.truncate()
+
+	s.state.statementStartLine = s.lineNum + 1
+
+	if err, ok := s.isDelimiterExpr(); err != nil {
+		s.err = err
+		return false
+	} else if ok {
+		// empty token acks DELIMITER
+		return true
+	}
+
+	for {
+		if err, ok := s.seekDelimiter(); err != nil {
+			s.err = err
+			return false
+		} else if ok {
+			// delimiter found, scanner holds valid token state
+			return true
+		} else if s.isEOF && s.i == s.fill {
+			// token terminates with file
+			s.state.end = s.fill
+			return true
+		}
+		// haven't found delimiter yet, keep reading
+		if err := s.read(); err != nil {
+			s.err = err
+			return false
+		}
+	}
+}
+
+func (s *streamScanner) truncate() {
+	// copy size should be 4k or less
+	copy(s.buf, s.buf[s.i:])
+	s.fill -= s.i
+	s.i = 0
+}
+
+func (s *streamScanner) resetState() {
+	s.state = &qState{}
+}
+
+func (s *streamScanner) read() error {
+	if len(s.buf) >= s.maxSize {
+		return fmt.Errorf("exceeded max query size")
+	}
+	newFillSize := s.fill + pageSize
+	if newFillSize > len(s.buf) {
+		newBufSize := min(len(s.buf)*2, s.maxSize)
+		newBuf := make([]byte, newBufSize)
+		copy(newBuf, s.buf)
+		s.buf = newBuf
+	}
+	// read at most |pageSize| into |s.buf| at index |s.fill|
+	n, err := s.inp.Read(s.buf[s.fill:newFillSize])
+	if err == io.EOF {
+		s.isEOF = true
+	} else if err != nil {
+		return err
+	}
+	// update fill, to |newFill| in the optimistic case
+	s.fill += n
+	return nil
+}
+
+func (s *streamScanner) Err() error {
+	return s.err
+}
+
+func (s *streamScanner) Bytes() []byte {
+	return s.buf[:s.state.end]
+}
+
+// Text returns the most recent token generated by a call to [Scanner.Scan]
+// as a newly allocated string holding its bytes.
+func (s *streamScanner) Text() string {
+	return string(s.buf[:s.state.end])
+}
+
+func (s *streamScanner) isDelimiterExpr() (error, bool) {
+	if s.i == 0 && s.fill-s.i < delimPrefixLen {
+		// need to see first 9 characters
+		if err := s.read(); err != nil {
+			s.err = err
+			return err, false
+		}
+	}
+
+	// valid delimiter state machine check
+	//  "DELIMITER " -> 0+ spaces -> <delimiter string> -> 1 space
+	if s.fill-s.i >= delimPrefixLen && bytes.EqualFold(s.buf[s.i:s.i+delimPrefixLen], delimPrefix) {
```

Comment:
are delimiters not case sensitive?

Reply:
i think case insensitive, `bytes.EqualFold`